### PR TITLE
Define types for plugin events

### LINKED ISSUE #6321: Add types for plugin events
### Current behavior:

Plugin events do not have TypeScript types.

### Desired behavior:

TypeScript types for plugin events

### LINKED ISSUE #6323: add JSDoc to scaffold plugins file
- we want to give users type-completions by adding a JSDoc comment to the scaffolded plugins file

this enables type completions for plugin events like `on('before:browser:launch')`

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #6321
- Closes #6323

### User facing changelog

Define TypeScript types for plugin events

### Additional details

This allows all the benefits of TypeScript for plugin events (type-checking, auto-completion, etc).

Documentation PR: https://github.com/cypress-io/cypress-documentation/pull/2466

### How has the user experience changed?

_I don't know how users can utilize this. I don't think it will work by default, so someone more knowledgeable with TypeScript will need to help answer that._

### PR Tasks
- [x] release new example-kitchen-sink with JSDoc comment in plugins file and bump version
fix #6323  
PR to add JSDoc in example-kitchensink https://github.com/cypress-io/cypress-example-kitchensink/pull/354

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by chrisbreiding** [comment]
@bahmutov How can we add tests for this [like for Cypress tests](https://github.com/cypress-io/cypress/tree/develop/cli/types/tests)? Is this even the right place to put this, next to the other types? They're 2 different environments...

File: cli/types/index.d.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ declare namespace Cypress {
   type RequestBody = string | object
   type ViewportOrientation = "portrait" | "landscape"
   type PrevSubject = "optional" | "element" | "document" | "window"
+  type PluginConfig = (on: PluginEvents, config: ConfigOptions) => void
```
Comment (bahmutov) [inline cli/types/index.d.ts]: the return type for plugin config function can return some props from config object or Promise to it or nothing I think

**Comment by jennifer-shehane** [comment]
I bumped the kitchensink dep with the released JSDoc changes.

**Comment by chrisbreiding** [comment]
@jennifer-shehane We should probably add this to the [TypeScript doc](https://docs.cypress.io/guides/tooling/typescript-support.html#Types-for-custom-commands), right? An example for how to utilize it is now in the [kitchen sink repo](https://github.com/cypress-io/cypress-example-kitchensink/blob/master/cypress/plugins/index.js)

**Comment by jennifer-shehane** [comment]
@chrisbreiding Yes, I'm not really familiar with what these changes mean for users, so open an issue/PR for docs.

**Comment by brian-mann** [comment]
@bahmutov and @flotwig and @Bkucera can ya'll comment here helping answer Chris's questions?

**Comment by chrisbreiding** [comment]
@brian-mann  My questions were answered. This is good to go now.

**Pull request merged**